src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -77,40 +88,56 @@ public boolean isRobotDistanceWithinRange() {
   }
 
   public double getDistanceFromGoal() {
-    //Todo: get distance from limelight
-    return SmartDashboard.getNumber("Distance from Goal", 0.0);
+    // distanceFromGoal is the following formula: (targetHeight - limelightHeight) / tan(limelightMountingAngle + limelightAngleToTarget)
+    return (Constants.kTargetHeight - Constants.kLimelightHeight) / 
+      Math.tan(Math.toRadians(Constants.kLimelightMountingAngle + mLimelightTable.getEntry("ty").getDouble(0.0))); 
+  }
+ 
+  public double getTargetX() {
+    return mLimelightTable.getEntry("tx").getDouble(0.0);
+  }
+
+  public double getDesiredTargetX() {
+    return mDesiredTargetX;
   }
 
-  public double getAngleFromGoal() {
-    //Todo: get angle from limelight
-    return SmartDashboard.getNumber("Angle from Goal", 0.0);
+  public boolean isTargetXAligned() {
+    return getTargetX() > mDesiredTargetX - Constants.kLimelightDrivebaseTolerance && 
+    getTargetX() < mDesiredTargetX + Constants.kLimelightDrivebaseTolerance;
   }
 
-  public boolean isAngleAligned() {
-    return getAngleFromGoal() == 0;
+  public void setLightStatus(boolean isOn) {
+    if (isOn) {
+      mLimelightTable.getEntry("ledMode").setNumber(Constants.kLedOn);
+    }
+    else {
+      mLimelightTable.getEntry("ledMode").setNumber(Constants.kLedOff);
+    }
   }
 
-  public boolean isReadyToShoot() {
-    return (isMotorVelocityWithinRange() && isAngleAligned());
+  public boolean isTargetInCameraFrame() {
+    return mLimelightTable.getEntry("tv").getNumber(0).doubleValue() > 0.0;
+  }
+
+  //Robot is close enough, shooter velocity is close enough, angle is close enough, target is on camera
+  public boolean isReadyToShoot() {    
+    return (isTargetInCameraFrame() && isMotorVelocityWithinRange() && isTargetXAligned() && isRobotDistanceWithinRange());
   }
 
   public double getRequiredVelocityForDistance() {
     double velocity; 
-    if (Constants.kGreenShootingZone < getDistanceFromGoal()) {
+    if (getDistanceFromGoal() < Constants.kGreenShootingZone) {
       velocity = mGreenShootingVelocity;
     } 
-    else if (Constants.kYellowShootingZone < getDistanceFromGoal()) {
+    else if (getDistanceFromGoal() < Constants.kYellowShootingZone) {
       velocity = mYellowShootingVelocity;
     } 
-    else if (Constants.kBlueShootingZone < getDistanceFromGoal()) {
+    else if (getDistanceFromGoal() < Constants.kBlueShootingZone) {
       velocity = mBlueShootingVelocity;
     }
-    else if (Constants.kRedShootingZone < getDistanceFromGoal()) {
+    else{ //Constants.kRedShootingZone < getDistanceFromGoal()
```

Comment:
Add comment "If we are outside of the other zones we can assume we are in the Red Zone"